File: L5/acoperireaConvexaAUnuiPoligonStelat2.py
# Graham’s scan, varianta Andrew (algoritm)

def determinant(xp,yp, xq,yq, xr,yr):
    det = 1*xq*yr + 1*xr*yp + 1*xp*yq - 1*xq*yp - 1*xr*yq - 1*xp*yr
    return det

def distantaLaPatrat(x1, y1, x2, y2):
    return (x1 - x2)**2 + (y1 - y2)**2

n = int(input())
P = []
k = 0
for i in range(n):
    xi, yi = map(int, input().split())
    P.append((xi, yi))

P.sort()

Li = []
Li.append(P[0])
Li.append(P[1])

for i in range(2, n):
    Li.append(P[i])
    while len(Li) > 2 and determinant(Li[-3][0], Li[-3][1], Li[-2][0], Li[-2][1], Li[-1][0], Li[-1][1]) <= 0:
        Li.pop(-2)


Ls = []
Ls.append(P[-1])
Ls.append(P[-2])

for i in range(n-3, -1, -1):
    Ls.append(P[i])
    while len(Ls) > 2 and determinant(Ls[-3][0], Ls[-3][1], Ls[-2][0], Ls[-2][1], Ls[-1][0], Ls[-1][1]) <= 0:
        Ls.pop(-2)

L = Li + Ls[1:-1]

print(len(L))
for Pi in L:
    print(Pi[0], Pi[1])

File: L5/acoperireaConvexaAUnuiPoligonStelat2.cpp
// Graham’s scan, varianta Andrew (algoritm)

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Punct {
    int x, y;
    bool operator<(const Punct& p) const {
        return (x < p.x) || (x == p.x && y < p.y);
    }
};

long long determinant(int xp, int yp, int xq, int yq, int xr, int yr) {
    return static_cast<long long>(xq) * yr + static_cast<long long>(xr) * yp + static_cast<long long>(xp) * yq
           - static_cast<long long>(xq) * yp - static_cast<long long>(xr) * yq - static_cast<long long>(xp) * yr;
}

int distantaLaPatrat(int x1, int y1, int x2, int y2) {
    return (x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2);
}

int main() {
    int n;
    cin >> n;
    vector<Punct> P(n);
    for (int i = 0; i < n; i++) {
        cin >> P[i].x >> P[i].y;
    }

    sort(P.begin(), P.end());

    vector<Punct> Li;
    Li.push_back(P[0]);
    Li.push_back(P[1]);

    for (int i = 2; i < n; i++) {
        Li.push_back(P[i]);
        while (Li.size() > 2 && determinant(Li[Li.size() - 3].x, Li[Li.size() - 3].y, Li[Li.size() - 2].x, Li[Li.size() - 2].y, Li[Li.size() - 1].x, Li[Li.size() - 1].y) <= 0) {
            Li.erase(Li.end() - 2);
        }
    }

    vector<Punct> Ls;
    Ls.push_back(P[n - 1]);
    Ls.push_back(P[n - 2]);

    for (int i = n - 3; i >= 0; i--) {
        Ls.push_back(P[i]);
        while (Ls.size() > 2 && determinant(Ls[Ls.size() - 3].x, Ls[Ls.size() - 3].y, Ls[Ls.size() - 2].x, Ls[Ls.size() - 2].y, Ls[Ls.size() - 1].x, Ls[Ls.size() - 1].y) <= 0) {
            Ls.erase(Ls.end() - 2);
        }
    }

    vector<Punct> L = Li;
    for (vector<Punct>::size_type i = 1; i < Ls.size() - 1; ++i) {
        L.push_back(Ls[i]);
    }

    cout << L.size() << endl;
    for (const auto& Punct : L) {
        cout << Punct.x << " " << Punct.y << endl;
    }

    return 0;
}
